README
 _____ ____  __  __           _____                 _       _
|_   _|  _ \|  \/  |         | ____|_ __ ___  _   _| | __ _| |_ ___  _ __
  | | | |_) | |\/| |  _____  |  _| | '_ ` _ \| | | | |/ _` | __/ _ \| '__|
  | | |  __/| |  | | |_____| | |___| | | | | | |_| | | (_| | || (_) | |
  |_| |_|   |_|  |_|         |_____|_| |_| |_|\__,_|_|\__,_|\__\___/|_|

 TPM-Emulator v0.7 -
 A Software-based TPM and MTM Emulator.
--------------------------------------------------------------------------

Copyright
--------------------------------------------------------------------------
Copyright (C) 2009-2011 Mario Strasser.
Copyright (C) 2004-2009 Mario Strasser, ETH Zurich.


This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.


Package
--------------------------------------------------------------------------
The tpm emulator package comprises four main parts:

a) tpmd - a user-space application that implements the actual emulator
   and can be accessed by means of Unix domain sockets (Unix) or
   named pipes (Windows).

b) tpmd_dev - a kernel module that provides the device /dev/tpm for
   backward compatibility and forwards the received commands to tpmd
   (Unix and Mac OS X only).

c) tddl - a TSS conform device driver library for the emulator.


Compilation and Installation
--------------------------------------------------------------------------
The compilation and installation of the TPM emulator package is based on
the CMake build environment (version 2.6 or better) and requires that the 
GNU MP library (version 4.0 or better) is properly installed on your
system. A working MinGW compiler suite is further required on Windows
(see http://www.mingw.org/). To compile and install the package execute:

# tar -xvzf tpm_emulator-X.Y.tar.gz
# cd tpm_emulator-X.Y
# mkdir build
# cd build
# cmake ../
# make
# make install

The script files build.sh and build.bat automate the compliation process.

On Windows, the TPM emulator system service has additionally to be
registered by calling:
# control_tpmd.bat install

MTM support can be enabled by replacing
# cmake ../
with
# cmake ../ -DMTM_EMULATOR=ON


Startup
--------------------------------------------------------------------------
In order to use the TPM emulator on Unix or Mac OS X, one has to start the
TPM emulator daemon and load the TPM device forwarding module. On Linux,
this is done by executing:

# modprobe tpmd_dev
# tpmd

On Windows, the TPM emulator service can either be started with the
Microsoft Management Console or with the control_tpmd.bat script:

# control_tpmd.bat start
# control_tpmd.bat status

The startup mode of the TPM (see TPM Spec. Part 1) is defined by the
startup mode argument and can either be set to clear, save (default)
or deactivated. Additionally supported arguments are

usage: tpmd [-d] [-f] [-s storage file] [-u unix socket name]
            [-o user name] [-g group name] [-h] [startup mode]
  d : enable debug mode
  f : forces the application to run in the foreground
  s : storage file to use (default: /var/lib/tpm/tpm_emulator-1_2_0_7)
  u : unix socket name to use (default: /var/run/tpm/tpmd_socket:0)
  o : effective user the application should run as
  g : effective group the application should run as
  h : print this help message
  startup mode : must be 'clear', 'save' (default) or 'deactivated

and

usage: tpmd.exe [-d] [-f] [-s storage file] [-u windows pipe name]
                [-l log file] [-h] [startup mode]
  d : enable debug mode
  f : forces the application to run in the foreground
  s : storage file to use (default:
      C:/Program Files/TPM_Emulator/tpm_emulator-1_2_0_7)
  u : windows named pipe name to use (default: //./pipe/tpmd:0)
  l : name of the log file (default: C:/Program Files/TPM_Emulator/tpmd.log)
  h : print this help message
  startup mode : must be 'clear', 'save' (default) or 'deactivated

on Unix and Windows, respectively.

If the emulator is started in mode save and fails to load a previously
stored TPM state, it will go into fail-stop mode and has to be reloaded.
Therefore, the first time the TPM emulator is started, the argument must
be set to 'clear'. Recovering a TPM emulator that is in fail-stop mode
is done by first deactivating it and then reloading it in mode 'clear':

# tpmd deactivated
# killall tpmd
# tpmd clear


Usage and Backward Compatibility
--------------------------------------------------------------------------
The most correct and convenient way to access the tpm emulator is to use
the provided device driver library (tddl). For a comprehensive description
of its functionality we refer to the official TCG specification (see
https://www.trustedcomputinggroup.org/specs/TSS), an example of use is
given by the test application tddl/test_tddl.

Note that on Windows the tddl is called ifxtpm.dll as many applications
(e.g., TPM/J) expect this name and do not support alternative drivers.

For backward compatibility with existing Unix applications, the kernel
module tpmd_dev forwards any command sent to the device /dev/tpm to the
tpm emulator daemon. In order to access the emulator directly (i.e.,
without using the device driver library or the device dev/tpm) all one
has to do is to include the header files sys/socket.h and sys/un.h and
to replace the open("/dev/tpm") call with something like:

struct sockaddr_un addr;
fh = socket(PF_UNIX, SOCK_STREAM, 0);
if (fh < 0) { /* handle error */ }
addr.sun_family = AF_UNIX;
strncpy(addr.sun_path, "/var/run/tpm/tpmd_socket:0", sizeof(addr.sun_path));
res = connect(fh, (struct sockaddr*)&addr, sizeof(struct sockaddr_un));
if (res < 0) { /* handle error */ }

All subsequent calls of read(), write(), and close() should work as
expected.


Usage Examples for TPM/J
--------------------------------------------------------------------------

=== Linux
# modprobe tpmd_dev
# tpmd -d

# cd <path to tpmj>/lib
# export CLASSPATH=tpmj.jar:bcprov-jdk15-131.jar:$CLASSPATH
# java edu.mit.csail.tpmj.tools.TPMInfo


=== Mac OS X
# sudo kextload /System/Library/Extensions/tpm_bridge.kext
# sudo tpmd -d

# cd <path to tpmj>/lib
# export CLASSPATH=tpmj.jar:bcprov-jdk15-131.jar:$CLASSPATH
# sudo java edu.mit.csail.tpmj.tools.TPMInfo


=== Windows
# set PATH=%PROGRAMFILES%\TPM_Emulator;%PATH%
# set PATH=%PROGRAMFILES%\TPM_Emulator\lib;%PATH%
# control_tpmd.bat start

# cd <path to tpmj>\lib
# set CLASSPATH=tpmj.jar;bcprov-jdk15-131.jar:%CLASSPATH%
# java edu.mit.csail.tpmj.tools.TPMInfo


Notes
--------------------------------------------------------------------------
* Since release 0.7 the emulator also runs on Darwin (Mac OS X) and on
  Windows; thanks go to Amit Singh and Domenic Schröder. In addition,
  the emulator now also supports MTM emulation; thanks go to Jan-Erik
  Ekberg and Markku Kylänpää from Nokia.
* Since release 0.5.1 the emulator supports Free- and OpenBSD; thanks go
  to Sebastian Schuetz.
* Since release 0.5 the kernel-based emulator (tpm_dev) is obsolete.
* The name and format of the persistent-storage file has changed between
  release 0.2 and 0.3, 0.4 and 0.4.1., 0.5 and 0.6, and 0.6 and 0.7.
* The DAA support was tested with the IBM DAA Test Suite and should work 
  now as expected. Thanks go to Roger Zimmermann for his kindly help.


Contact
--------------------------------------------------------------------------
Any comments, suggestions and bug reports are welcome. Please, mention
the keyword 'TPM emulator' in the subject.
Peter Huewe <peterhuewe@gmx.de>
Mario Strasser <mast@gmx.net>



ChangeLog
2011-11-05  Mario Strasser <mast@gmx.net>
	* ChangeLog: ChangeLog updated
	* CMakeList.txt: minor version number updated
	* tpm_integrity.c: Fixed bug in TPM_Quote2; thanks go to Andreas Steffen


2011-11-12  Mario Strasser <mast@gmx.net>
	* tpmd_dev/darwin/*: Added support for OS X 10.6; thanks go to Uri Blumenthal
	* tpm_counter.c: Bug in TPM_IncrementCounter fixed; thanks go to Charles T


2011-11-05  Mario Strasser <mast@gmx.net>
	* ChangeLog: ChangeLog updated
	* CMakeList.txt: minor version number updated


2011-10-25  Mario Strasser <mast@gmx.net>
	* tpm_owner.c: set Pubek to false in TakeOwnership


2011-09-15  Mario Strasser <mast@gmx.net>
	* tpmd_dev.c: fixed byter order conversion bug for aligned memory


2011-09-09  Mario Strasser <mast@gmx.net>
	* tpm_identity.c: bug fix for MakeIdentity;
		thanks go to Matthew Fioravante


2011-07-19  Mario Strasser <mast@gmx.net>
	* tpm_authorization.c, tpm_cmd_handler.c, tpm_structures.h:
		copy secret to auth struct


2011-07-19  Mario Strasser <mast@gmx.net>
	* tpm_owner.c: error codes revised


2011-07-19  Mario Strasser <mast@gmx.net>
	* tpm_nv_storage.c: bug in NV_DefineSpace fixed; thaks go to Luigi


2011-07-09  Mario Strasser <mast@gmx.net>
	* CMakeLists.txt, crypto/*: preliminary openssl support added


2011-06-08  Mario Strasser <mast@gmx.net>
	* tpm_emulator.h, tpm_owner.c, tpmd.c:
		removed TPM_CONF_KEEP_PUBEK_READABLE


2011-06-04  Mario Strasser <mast@gmx.net>
	* tpm_migration.c: bug in TPM_CMK_ApproveMA fixed


2011-02-13  Mario Strasser <mast@gmx.net>
	* CMakeLists.txt: minor package version updated


2011-02-12  Mario Strasser <mast@gmx.net>
	* ChangeLog: ChangeLog updated
	* CMakeLists.txt: minor package version updated
	* tpmd_dev.c: switched to new ioctl function


2010-12-30  Mario Strasser <mast@gmx.net>
	* build.bat: Missing path for Windows 7 added


2010-10-09  Mario Strasser <mast@gmx.net>
	* tpm_capability.c: Bug in TPM_SetCapability fixed


2010-09-11  Mario Strasser <mast@gmx.net>
	* all: Open/FreeBSD compatibility fixes


2010-07-19  Mario Strasser <mast@gmx.net>
	* tpm_daa.c: Bug in TPM_LoadContext and TPM_DAA_Join/Sign
		fixed; thanks go toEmanuele Cesena


2010-07-05  Mario Strasser <mast@gmx.net>
	* ChangeLog: ChangeLog updated
	* tpmd_dev/linux/*: missing kernel header file added


2010-06-13  Mario Strasser <mast@gmx.net>
	* tpm_migration.c, tpm_cmd_handler.c, tpm_crypto.c:
		bugs in [CMK]MigrationBlob encoding, TPM_MigrateKey(),
		and TPM_CertifyKey2 fixed; thanks go to S. Tate
	* tpm_structures.h: bug in TPM_CERTIFY_INFO fixed;
		thanks go to S. Tate


2010-06-10  Mario Strasser <mast@gmx.net>
	* tpm_nv_storage.c: bug in TPM_NV_WriteValue() fixed;
		thanks go to Luigi Semenzato


2010-03-17  Mario Strasser <mast@gmx.net>
	* tpm_emulator_extern.[ch]: external functions changed
		into function pointers
	* tpmd/*/tpmd.c: tpmd functions updated


2010-03-12  Mario Strasser <mast@gmx.net>
	* tpm_nv_storage.c: bug in TPM_NV_DefineSpace fixed;
		thanks go to Luigi Semenzato
	* README, build.sh: build description updated


2010-02-22  Mario Strasser <mast@gmx.net>
	* tpmd_dev/darwin/*: darwin module added
	* Makefile: installation on Mac OS X enabled
	* README: README updated
	* ChangeLog: ChangeLog updated


2010-02-19  Mario Strasser <mast@gmx.net>
	* windows/*: several fixes
	* crypto/bn.c: work around for older gmp libs
	* tddl/*: platform independent tddl
	* README: README updated


2010-02-18  Mario Strasser <mast@gmx.net>
	* build.sh, buils.bat: build scripts added
	* tpmd.c, tpm_emulator_extern.h: log modes independent of syslog
	* windows/tpmd.c: tpmd for windows added
	* windows/control_tpmd.bat: service control function added


2010-02-16  Mario Strasser <mast@gmx.net>
	* tpm/*: MTM functions added
	* tpmd.c: parameter to set configuration flags added
	* README: README updated


2010-02-15  Mario Strasser <mast@gmx.net>
	* tpm_emulator.h, tpm_cmd_handler.c: startup configuration added
	* CMakeLists.txt: build of openbsd module added
	* tpm/*: MTM support prepared and MTM hooks added


2010-02-13  Mario Strasser <mast@gmx.net>
	* all: unused variables removed; free_TPM_DATA added


2010-02-12  Mario Strasser <mast@gmx.net>
	* tpm_marshalling.[ch]: (un)marshaling_TPM_DATA added


2010-02-11  Mario Strasser <mast@gmx.net>
	* CMakeLists.txt: installation support added
	* README: README updated
	* *.h: headers updated


2010-02-10  Mario Strasser <mast@gmx.net>
	* all: new CMake makefiles added
	* README: README updated
	* CMakeLists.txt, config.h.in: cpack support added


2010-02-09  Mario Strasser <mast@gmx.net>
	* crypto/*, tpm/*: BE* conversion functions removed
	* tpmd/tpmd.c:  smart startup mode handling
	* crypto/*.h, tpm/*.h: tpm_emulator_config.h removed
	* tpm_testinc.c: usage of tpm_min/max removed
	* tpm_data.c, tpm_emulator_extern.h, tpmd.c:
		_file -> _storage
	* config.h: central config.h added
	* all: directory structure changed


2009-12-05  Mario Strasser <mast@gmx.net>
	* tpmd_dev/Makefile: installation patch added, thanks go to David Anderson


2009-10-05  Mario Strasser <mast@gmx.net>
	* all: numerous bug fixes to make the trousers test pass


2009-10-02  Mario Strasser <mast@gmx.net>
	* tpm_delegation.c: delegation support completed


2009-09-24  Mario Strasser <mast@gmx.net>
	* tpm_crypto.c: internal, deterministic PRNG added (for testing etc.)
	* tpm_authorization.c: DSAP verification added


2009-09-19  Mario Strasser <mast@gmx.net>
	* tpm_capability.c: TPM_SetCapability() added
	* tpm_capability.c: TPM_GetCapability() completed
	* ChangeLog: changelog updated


2009-09-18  Mario Strasser <mast@gmx.net>
	* tpm_commands.h: declarations and descriptions of internal
		functions added


2009-09-17  Mario Strasser <mast@gmx.net>
	* tpm_authorization.c: DSAP support added


2009-08-28  Mario Strasser <mast@gmx.net>
	* tpm_migration.c: CMK handling completed
	* tpm_owner.c: TPM_OwnerClear() updated
	* all: code cleanup, obsolete code removed

crypto/sha1.c
/* Software-based Trusted Platform Module (TPM) Emulator
 * Copyright (C) 2004-2010 Mario Strasser <mast@gmx.net>
 *
 * This module is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published
 * by the Free Software Foundation; either version 2 of the License,
 * or (at your option) any later version.
 *
 * This module is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * $Id: sha1.c 364 2010-02-11 10:24:45Z mast $
 */

#include "sha1.h"
#include <string.h>

/* This code is based on Steve Reid's <steve@edmweb.com> 
   public domain implementation. */ 

#define rol(v,b) (((v) << (b)) | ((v) >> (32 - (b))))
#ifdef __BIG_ENDIAN__
#define B0(i) (buf[i] = buf[i])
#else
#define B0(i) (buf[i] = (((buf[i] & 0xff000000) >> 24) \
                       | ((buf[i] & 0x00ff0000) >> 8) \
                       | ((buf[i] & 0x0000ff00) << 8) \
                       | ((buf[i] & 0x000000ff) << 24)))

#endif
#define B1(i) (buf[i & 15] = rol(buf[i & 15] ^ buf[(i-14) & 15] \
                                 ^ buf[(i-8) & 15] ^ buf[(i-3) & 15], 1))
#define F0(x,y,z) ((x & (y ^ z)) ^ z)
#define F1(x,y,z) (x ^ y ^ z)
#define F2(x,y,z) (((x | y) & z) | (x & y))
#define R0(a,b,c,d,e,i) e += F0(b,c,d) + B0(i) + 0x5A827999 + rol(a,5); b = rol(b,30);
#define R1(a,b,c,d,e,i) e += F0(b,c,d) + B1(i) + 0x5A827999 + rol(a,5); b = rol(b,30);
#define R2(a,b,c,d,e,i) e += F1(b,c,d) + B1(i) + 0x6ED9EBA1 + rol(a,5); b = rol(b,30);
#define R3(a,b,c,d,e,i) e += F2(b,c,d) + B1(i) + 0x8F1BBCDC + rol(a,5); b = rol(b,30);
#define R4(a,b,c,d,e,i) e += F1(b,c,d) + B1(i) + 0xCA62C1D6 + rol(a,5); b = rol(b,30);

static void tpm_sha1_transform(uint32_t h[5], const uint8_t data[64])
{
  uint32_t a, b, c, d, e;
  uint32_t buf[16];

  /* copy state and data*/
  a = h[0];
  b = h[1];
  c = h[2];
  d = h[3];
  e = h[4];
  memcpy(buf, data, 64);
  /* unrolled sha-1 rounds */
  R0(a,b,c,d,e, 0); R0(e,a,b,c,d, 1); R0(d,e,a,b,c, 2); R0(c,d,e,a,b, 3);
  R0(b,c,d,e,a, 4); R0(a,b,c,d,e, 5); R0(e,a,b,c,d, 6); R0(d,e,a,b,c, 7);
  R0(c,d,e,a,b, 8); R0(b,c,d,e,a, 9); R0(a,b,c,d,e,10); R0(e,a,b,c,d,11);
  R0(d,e,a,b,c,12); R0(c,d,e,a,b,13); R0(b,c,d,e,a,14); R0(a,b,c,d,e,15);
  R1(e,a,b,c,d,16); R1(d,e,a,b,c,17); R1(c,d,e,a,b,18); R1(b,c,d,e,a,19);
  R2(a,b,c,d,e,20); R2(e,a,b,c,d,21); R2(d,e,a,b,c,22); R2(c,d,e,a,b,23);
  R2(b,c,d,e,a,24); R2(a,b,c,d,e,25); R2(e,a,b,c,d,26); R2(d,e,a,b,c,27);
  R2(c,d,e,a,b,28); R2(b,c,d,e,a,29); R2(a,b,c,d,e,30); R2(e,a,b,c,d,31);
  R2(d,e,a,b,c,32); R2(c,d,e,a,b,33); R2(b,c,d,e,a,34); R2(a,b,c,d,e,35);
  R2(e,a,b,c,d,36); R2(d,e,a,b,c,37); R2(c,d,e,a,b,38); R2(b,c,d,e,a,39);
  R3(a,b,c,d,e,40); R3(e,a,b,c,d,41); R3(d,e,a,b,c,42); R3(c,d,e,a,b,43);
  R3(b,c,d,e,a,44); R3(a,b,c,d,e,45); R3(e,a,b,c,d,46); R3(d,e,a,b,c,47);
  R3(c,d,e,a,b,48); R3(b,c,d,e,a,49); R3(a,b,c,d,e,50); R3(e,a,b,c,d,51);
  R3(d,e,a,b,c,52); R3(c,d,e,a,b,53); R3(b,c,d,e,a,54); R3(a,b,c,d,e,55);
  R3(e,a,b,c,d,56); R3(d,e,a,b,c,57); R3(c,d,e,a,b,58); R3(b,c,d,e,a,59);
  R4(a,b,c,d,e,60); R4(e,a,b,c,d,61); R4(d,e,a,b,c,62); R4(c,d,e,a,b,63);
  R4(b,c,d,e,a,64); R4(a,b,c,d,e,65); R4(e,a,b,c,d,66); R4(d,e,a,b,c,67);
  R4(c,d,e,a,b,68); R4(b,c,d,e,a,69); R4(a,b,c,d,e,70); R4(e,a,b,c,d,71);
  R4(d,e,a,b,c,72); R4(c,d,e,a,b,73); R4(b,c,d,e,a,74); R4(a,b,c,d,e,75);
  R4(e,a,b,c,d,76); R4(d,e,a,b,c,77); R4(c,d,e,a,b,78); R4(b,c,d,e,a,79);
  /* update state */
  h[0] += a;
  h[1] += b;
  h[2] += c;
  h[3] += d;
  h[4] += e;
  /* overwrite all used variables */
  a = b = c = d = e = 0;
  memset(buf, 0, 64);
}


void tpm_sha1_init(tpm_sha1_ctx_t *ctx)
{
  /* initialise with sha-1 constants */
  ctx->h[0] = 0x67452301;
  ctx->h[1] = 0xEFCDAB89;
  ctx->h[2] = 0x98BADCFE;
  ctx->h[3] = 0x10325476;
  ctx->h[4] = 0xC3D2E1F0;
  ctx->count_lo = ctx->count_hi = 0;
}

void tpm_sha1_update(tpm_sha1_ctx_t *ctx, const uint8_t *data, size_t length)
{
  size_t buf_off = (ctx->count_lo >> 3) & 63;
  size_t data_off = 0;

  /* add data */
  if (length + buf_off >= 64) {
    data_off = 64 - buf_off;
    memcpy(&ctx->buf[buf_off], data, data_off);
    tpm_sha1_transform(ctx->h, ctx->buf);
    while (data_off + 64 <= length) { 
      tpm_sha1_transform(ctx->h, &data[data_off]);
      data_off += 64;
    }
    buf_off = 0;
  }
  memcpy(&ctx->buf[buf_off], &data[data_off], length - data_off);
  /* update counter */
  buf_off = ctx->count_lo;
  ctx->count_lo += length << 3;
  if (ctx->count_lo < buf_off) ctx->count_hi++;
  ctx->count_hi += length >> 29;
}

void tpm_sha1_update_be32(tpm_sha1_ctx_t *ctx, uint32_t data)
{
  uint8_t buf[4];

  buf[0] = (data >> 24) & 0xff;
  buf[1] = (data >> 16) & 0xff;
  buf[2] = (data >>  8) & 0xff;
  buf[3] = (data >>  0) & 0xff;
  tpm_sha1_update(ctx, buf, 4);
}

void tpm_sha1_final(tpm_sha1_ctx_t *ctx, uint8_t digest[SHA1_DIGEST_LENGTH])
{
  uint8_t d, counter[8];

  /* setup counter */
  for (d = 0; d < 4; d++) {
    counter[d    ] = (ctx->count_hi >> (24 - d * 8)) & 0xff;
    counter[d + 4] = (ctx->count_lo >> (24 - d * 8)) & 0xff;
  }
  /* add padding */
  d = 0x80;
  tpm_sha1_update(ctx, &d, 1);
  d = 0x00;
  while ((ctx->count_lo & (63 * 8)) != (56 * 8)) tpm_sha1_update(ctx, &d, 1);
  /* add counter */
  tpm_sha1_update(ctx, counter, 8);
  for (d = 0; d < SHA1_DIGEST_LENGTH; d++) 
    digest[d] = (uint8_t)(ctx->h[d >> 2] >> (8 * (3 - (d & 3))) & 0xff);
  /* overwrite all used variables */
  memset(ctx, 0, sizeof(*ctx));
  memset(counter, 0, sizeof(counter));
}
